replica/database.cc

```diff
@@ -2571,21 +2571,37 @@ future<> database::clear_snapshot(sstring tag, std::vector<sstring> keyspace_nam
                         if (tag.empty()) {
                             dblog.info("Removing {}", snapshots_dir);
                             recursive_remove_directory(std::move(snapshots_dir)).get();
+                            has_snapshots = false;
                         } else {
                             // if specific snapshots tags were given - filter only these snapshot directories
-                            auto snapshots_dir_lister = directory_lister(snapshots_dir, {directory_entry_type::directory},
-                                    [&] (const fs::path&, const directory_entry& dir_entry) { return dir_entry.name == tag; });
+                            auto snapshots_dir_lister = directory_lister(snapshots_dir, {directory_entry_type::directory});
                             auto close_snapshots_dir_lister = deferred_close(snapshots_dir_lister);
                             dblog.debug("clear_snapshot: listing snapshots dir {} with filter={}", snapshots_dir, tag);
+                            has_snapshots = false;  // unless other snapshots are found
                             while (auto snapshot_ent = snapshots_dir_lister.get().get0()) {
-                                auto snapshot_dir = snapshots_dir / snapshot_ent->name;
-                                dblog.info("Removing {}", snapshot_dir);
-                                recursive_remove_directory(std::move(snapshot_dir)).get();
+                                if (snapshot_ent->name == tag) {
+                                    auto snapshot_dir = snapshots_dir / snapshot_ent->name;
+                                    dblog.info("Removing {}", snapshot_dir);
+                                    recursive_remove_directory(std::move(snapshot_dir)).get();
+                                } else {
+                                    has_snapshots = true;
+                                }
                             }
                         }
                     } else {
                         dblog.debug("clear_snapshot: {} not found", snapshots_dir);
                     }
+                    // zap the table directory if the table is dropped
+                    // and has no remaining snapshots
+                    if (!has_snapshots) {
+                        auto [cf_name, cf_uuid] = extract_cf_name_and_uuid(table_ent->name);
+                        const auto& it = _ks_cf_to_uuid.find(std::make_pair(ks_name, cf_name));
+                        auto dropped = (it == _ks_cf_to_uuid.cend()) || (cf_uuid != it->second);
+                        if (dropped) {
+                            dblog.info("Removing dropped table dir {}", table_dir);
+                            recursive_remove_directory(table_dir).get();
```

Comment:
I can't find any bug in this code, but it looks scary - if the logic above is somewhat wrong, you can find yourself removing an entire non-empty directory. If table_dir is somehow (?) "/", you find yourself removing the entire disk ;-)
Again, judicious use of unlink() to remove only guaranteed-empty directories seems safer to me, although I admit there's nothing wrong in your code that I can find.

Reply:
No problem.  I'll give the safer unlink approach a shot.
Although I will use `recursive_remove_directory` for table subdirectories, but this is safe than removing just `table_dir`.
It should never ever become "/" or empty, but I guess a bug in the future might cause that, and adding a subdirectory to it will prevent/limit the damage.

Reply:
I thought (but maybe I'm missing something) that you can assume that at this stage, there would only be empty subdirectories - there will be an empty "staging" subdirectory (because you said it's removed earlier), an empty "snapshots" subdirectory (if there are no snapshots), etc. So you can use unlink() for all of them. If unlink("snapshots") fails, it means there are snapshots, and you don't even need to check.
What subdirectory can be non-empty and require recursive_remove_directory?

Reply:
Normally, the sub-directories would be empty, but I really don't think we should care so much about it.
Otherwise, we won't be able to get rid of stale directories of dropped tables that happened to have some unexpected junk in them (from the point of view of a pedantic function).  For example, a table that was removed using an older version of scylla, or a table we want to drop after an error (e.g. ENOSPC) that might have stuff under the "pending_delete" subdirectory, that can't proceed for some reason.

Bottom line, is I think that making this function pedantic might require too much code and I don't see the benefit in that.
If the user asked to drop the table we should remove everything under the table directory other than live snapshots.

We may need special care for snapshots in the future though since I'm considering a change in the snapshot taking logic to create the snapshot in a temporary directory and move it to its final name only when taking the snapshot is successfully done, to prevent cases where we end up with partially built snapshot after an error.  Those temporary snapshot directories should be cleaned up after the error, and we can clean them up on startup, but I think we should also delete them unconditionally when removing the table dir.